Backport AD4000 series driver with offload support

machschmitt opened the pull request:
## PR Description

Ports back the AD4000 driver into ADI Linux with offload support as requested at https://baylibre.atlassian.net/browse/ADI-369.

0) Fixes IIO doc index.
1) Backports `devm_mutex_init()` API form IIO tree.
2) Cherry-picks upstream commits that conflicted with SPI MOSI idle feature.
3)  Ports the SPI MOSI idle support from upstream into ADI Linux.
4) Removes the ad400x driver
5) Add device tree doc for AD4000 series
6) Updates existing device tree of AD4000 series devices
7) Adds the ad4000 driver from upstream
8) Backport a couple of upstream fixes for the ad4000 driver.
9) Update ADI imply for AD4000
10) Add support for adjustable sampling rate
11) Implements timestamp channel

Some commits for the dts files are very similar and may be squashed. I decided to leave them separate so it's more clear what the changes are. Also, squashing is easier than splitting.

Previously, the ad4000 driver was adapted to avoid cherry-picking additional commits from upstream. With recently backported features to ADI main plus the `devm_mutex_init()` API, AD4000 series support can be backported without any changes to driver code.

Tested with ADAQ4003 on rpi and on cora.
Plan is to support single-channel pulsar (from ad_pulsar.c driver) in ad4000.c in the future.

## PR Type
- [x] Bug fix (a change that fixes an issue)
- [x] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware
- [ ] I have updated the documentation outside this repo accordingly (if there is the case)

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -349,6 +375,52 @@ static int ad4000_single_conversion(struct iio_dev *indio_dev,
+	 * 	⟺ P ≥ round_up(R / freq) * NSEC_PER_SEC / R
+	 */
+
+	cnv_state.period = div_u64_rem((u64)DIV_ROUND_UP(st->ref_clk_rate, freq) * NSEC_PER_SEC,
```
FYI, @ukleinek is working on backporting some PWM changes so we will have a better way to do this. 

This is probably fine for now, but one to add to the todo list to fix up later.

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -477,6 +571,68 @@ static const struct iio_info ad4000_info = {
+	int ret;
+
+	spi_bus_lock(st->spi->master);
+	st->bus_locked = true;
```
This variable is set but never read, so probably can be removed.

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -211,13 +232,18 @@ static const struct ad4000_chip_info adaq4003_chip_info = {
+	bool bus_locked;
+	unsigned long ref_clk_rate;
+	struct pwm_device *cnv_trigger;
+	int max_rate;
```
Upstream prefers units in variable names `ref_clk_rate_hz`, `max_rate_hz`.

dlech commented:
Answering question from the last commit message:

> Use multiple channel scan_type to have different buffer endianness for each buffer setup?

Yes. Jonathan's recommendation on similar upstream patches is to make a separate channel info array for the offload version vs. the not-offload.

So in addition to the different endianness, the `IIO_CHAN_INFO_SAMP_FREQ` flag should also only be present in this new offload channel info since it isn't present when there isn't an offload. Then we can remove the `return -ENOTSUPP;` in the raw_read and raw_write callbacks since the core iio code won't create the attributes in that case.

dlech commented:
Also, FYI, a bunch of regulator and other dependencies are being backported in https://github.com/analogdevicesinc/linux/pull/2433, so we should be able to make use of those and avoid some (all?) of the "Adapt to not use..."  patches here.

machschmitt commented:
Hi @dlech 

> So in addition to the different endianness, the `IIO_CHAN_INFO_SAMP_FREQ` flag should also only be present in this new offload channel info since it isn't present when there isn't an offload. Then we can remove the `return -ENOTSUPP;` in the raw_read and raw_write callbacks since the core iio code won't create the attributes in that case.

Yes, that would be ideal. Though, the `iio_chan_spec` structs are all statically declared and we don't have the info on whether offloading is supported at build time. So, to have `IIO_CHAN_INFO_SAMP_FREQ` only present when offload is supported, it would require either duplicating the statically declared `iio_chan_spec` structs (half for non-offload, half for offload) or populating the `iio_chan_spec` only when the device probes. I'm not very excited about the `iio_chan_spec` explosion so the latter option looks better to me. Yet, scraping the statically declared `iio_chan_spec` structs (in favor of runtime-populated chan_spec) will lead to some driver rework I was trying to avoid. So maybe this will become more common to drivers that support offloading: to have IIO channel structs populated at runtime as opposed to statically declared ones. Anyway, appreciate any comments about it. Thanks.

dlech commented:
> it would require either duplicating the statically declared `iio_chan_spec` structs

I can't find the thread at the moment to quote Jonathan Cameron exactly, but duplicating the static const structs is the preferred solution upstream even it is seems a bit overkill. There are already a number of other drivers doing something similar.

machschmitt commented:
Change log v1 -> v2:
- Rebased on top of ADI main.
- Ported devm_mutex_init() <4cd47222e435> ("locking/mutex: Introduce devm_mutex_init()") to ADI Linux.
- Updated ad4000 driver to take advantage of useful API recently added to ADI main. Now the only differences with respect to upstream ad4000 driver are related to offload support.
- Updated to have different statically declared `iio_chan_info` structs for when SPI offloading is available.
- Added separated `iio_info` struct for offload support so attribute write is not allowed if reg access and offloading are not supported.

machschmitt commented on drivers/iio/adc/ad4000.c:
```diff
@@ -656,11 +862,28 @@ static int ad4000_probe(struct spi_device *spi)
-	if (ret)
-		return ret;
+	if (spi_engine_ex_offload_supported(spi)) {
+		if (device_property_present(&spi->dev, "pwms")) {
```
Currently, no. The PWM is needed. Okay, will hard fail if no pwm is available.

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -494,6 +681,7 @@ static int ad4000_prepare_3wire_mode_message(struct ad4000_state *st,
 
 	xfers[1].rx_buf = &st->scan.data;
+	xfers[1].bits_per_word = chan->scan_type.storagebits;
 	xfers[1].len = BITS_TO_BYTES(chan->scan_type.storagebits);
```
Why not? It worked for me on ad7985 with SPI offload which should be very similar.

machschmitt replied:
I don't know why, but the capture looks like this if 
```
-	xfers[1].bits_per_word = chan->scan_type.storagebits;
+	xfers[1].bits_per_word = chan->scan_type.realbits;
```
![Screenshot at 2024-10-29 17-31-42](https://github.com/user-attachments/assets/dd46f875-c3e0-4bb4-932e-3c24d46e0b1b)
I can set it up on rpi to check whether triggered buffer breaks or not with `xfers[1].bits_per_word = chan->scan_type.storagebits;`. Anyway, realbits doesn't seem to work with offload. Can it be different HDL?

dlech replied:
> I can set it up on rpi to check whether triggered buffer breaks or not with xfers[1].bits_per_word = chan->scan_type.storagebits;

I would expect this to fail on RPI because it can only do 8 bits per word.

> Can it be different HDL?

That was my first suspicion. What is the `realbits` of the chip you are testing with?

machschmitt replied:
> That was my first suspicion. What is the `realbits` of the chip you are testing with?

The ADC I have tested this with is ADAQ4003 (18-bit precision) (CoraZ7 setup).
I have now tried somewhat recent HDL from ADI artifactory (https://artifactory.analog.com/ui/repos/tree/General/sdg-generic-development%2Flinux%2Freleases%2Fnext_stable%2F2024_09_26-06_39_10) but that fails due to lack of the MOSI idle feature (`ad4000 spi0.0: setup: unsupported mode bits 40000`).
Well, would we have a 18-bit words (or maybe 24-bit words) with `bits_per_word = realbits`? Wouldn't that mess up with the byte swapping when the CPU translates the data to its endianness?
The BOOT.BIN I was using during the tests yesterday was this one 
[2024-08-08-cora-pulsar_adc-main-debug_v2-BOOT.BIN.zip](https://github.com/user-attachments/files/17571516/2024-08-08-cora-pulsar_adc-main-debug_v2-BOOT.BIN.zip)
which I built from HDL main branch with some modifications to insert ILA cores (debug).

machschmitt replied:
With 
```c
	if (chan->scan_type.endianness != IIO_BE)
		xfers[1].bits_per_word = chan->scan_type.realbits;
```
the transfer `bits_per_word` field is not set and gets a default of 8 bits per in spi_setup() 
```c
	if (!spi->bits_per_word) {
		spi->bits_per_word = 8;
```
The SPI subsystem runs the transfers and pushes data to memory in 8-bit wordsizes,
preserving the same data endianness as the peripheral device. Latter, the transfer
data can be read from memory by IIO drivers or by the IIO buffer handling routines
which we program to handle the data endianness according to device data output
endianness. For single-shot read, `be16_to_cpu()` is often used if the data
is not addressed through an array. For buffer configuration we have IIO_BE,
IIO_LE, and IIO_CPU, being IIO_BE what we often use since many ADCs output data
in big endian format. Thus, with `bits_per_word` unset (8-bit words) everything
works as expected because the endianness handling is done by IIO only.

If `xfers[1].bits_per_word = chan->scan_type.storagebits;` is set unconditionally
as in previous ad4000 version, the transfer data is stored in memory in
"in-memory wordsizes" of 32-bits per word and the data is translated to CPU
endianness at some point (either by the SPI subsystem or by some other part of
the kernel). Though, when the IIO driver latter access the transfer data from
memory it would miss handle the data thinking it is in BE while it would be in
CPU endianness (LE for our arm carrier boards). So,
`xfers[1].bits_per_word = chan->scan_type.storagebits;` does indeed break the
previous support for rpi. On the other hand, if `bits_per_word` is set to 32
and the buffer endianness is changed to IIO_CPU, it works just like when
`bits_per_word` was not set.

The lesson to be learned here is that we have some
flexibility in where data endianness is handled. We can set proper
`bits_per_word` in transfers to make the underlying kernel subsystems handle
endianness, or we can choose to handle data endianness ourselves in IIO.

machschmitt commented:
Change log v2 -> v3
- Swapped PWM devm_add_action_or_reset() with PWM set so the PWM state is set prior to calling action_or_reset().
- Fixed iio_info choice when SDI_MOSI
- Hard fail if PWM and related clk are not available in offload mode
- Always use 32 storagebits if offload is supported
- Switch _offl / _reg_access order
-  Use MEGA and KILO to set `max_rate_hz`

dlech commented:
I dug into this deeper and actually tested it today with a logic analyzer and found a few issues. I didn't have the right eval board to test with, so this might not be 100% correct, but at least a step in the right direction (I was able to test the non-turbo mode with AD7985 since it has the same wiring as `adi,sdi-pin = "high";`).

The main issue is that we need very specialized SPI messages in order to actually get good samples at the max sample rate. What the max sample rate is and timings of the transfer depend on if TURBO is enabled or not. So this is going to depend on if we can write registers or not. And the SPI message for single conversions isn't going to work because that does two conversions, which takes too long at high sample rates.

Here is a diff of the changes I made on to of this series to get things at least sort of working:

```diff
diff --git a/drivers/iio/adc/ad4000.c b/drivers/iio/adc/ad4000.c
index 56cfff16ea54..071cf5a3673f 100644
--- a/drivers/iio/adc/ad4000.c
+++ b/drivers/iio/adc/ad4000.c
@@ -59,7 +59,7 @@
 		.sign = _sign,							\
 		.realbits = _real_bits,						\
 		.storagebits = _storage_bits,					\
-		.shift = _storage_bits - _real_bits,				\
+		.shift = _offl ? 0 :_storage_bits - _real_bits,			\
 		.endianness = _offl ? IIO_CPU : IIO_BE				\
 	},									\
 }
@@ -83,7 +83,7 @@
 		.sign = _sign,							\
 		.realbits = _real_bits,						\
 		.storagebits = _storage_bits,					\
-		.shift = _storage_bits - _real_bits,				\
+		.shift = _offl ? 0 :_storage_bits - _real_bits,			\
 		.endianness = _offl ? IIO_CPU : IIO_BE				\
 	},									\
 }
@@ -278,6 +278,8 @@ struct ad4000_state {
 	struct gpio_desc *cnv_gpio;
 	struct spi_transfer xfers[2];
 	struct spi_message msg;
+	struct spi_transfer offload_xfers[2];
+	struct spi_message offload_msg;
 	unsigned long ref_clk_rate_hz;
 	struct pwm_device *cnv_trigger;
 	int max_rate_hz;
@@ -294,8 +296,10 @@ struct ad4000_state {
 	 */
 	struct {
 		union {
-			__be16 sample_buf16;
-			__be32 sample_buf32;
+			__be16 sample_buf16_be;
+			__be32 sample_buf32_be;
+			u16 sample_buf16;
+			u32 sample_buf32;
 		} data;
 		s64 timestamp __aligned(8);
 	} scan __aligned(IIO_DMA_MINALIGN);
@@ -398,15 +402,24 @@ static int ad4000_single_conversion(struct iio_dev *indio_dev,
 	if (ret < 0)
 		return ret;
 
-	if (chan->scan_type.storagebits > 16)
-		sample = be32_to_cpu(st->scan.data.sample_buf32);
-	else
-		sample = be16_to_cpu(st->scan.data.sample_buf16);
+	if (chan->scan_type.endianness == IIO_BE) {
+		if (chan->scan_type.realbits > 16)
+			sample = be32_to_cpu(st->scan.data.sample_buf32_be);
+		else
+			sample = be16_to_cpu(st->scan.data.sample_buf16_be);
+	} else {
+		if (chan->scan_type.realbits > 16)
+			sample = st->scan.data.sample_buf32;
+		else
+			sample = st->scan.data.sample_buf16;
+	}
 
 	sample >>= chan->scan_type.shift;
 
 	if (chan->scan_type.sign == 's')
 		*val = sign_extend32(sample, chan->scan_type.realbits - 1);
+	else
+		*val = sample;
 
 	return IIO_VAL_INT;
 }
@@ -599,8 +612,10 @@ static int ad4000_buffer_postenable(struct iio_dev *indio_dev)
 	struct ad4000_state *st = iio_priv(indio_dev);
 	int ret;
 
+	/* TODO: need to enable TURBO mode via registers, if applicable. */
+
 	spi_bus_lock(st->spi->master);
-	ret = spi_engine_ex_offload_load_msg(st->spi, &st->msg);
+	ret = spi_engine_ex_offload_load_msg(st->spi, &st->offload_msg);
 	if (ret < 0)
 		return ret;
 
@@ -618,6 +633,8 @@ static int ad4000_buffer_postdisable(struct iio_dev *indio_dev)
 
 	spi_bus_unlock(st->spi->master);
 
+	/* TODO: need to disable TURBO mode via registers, if applicable. */
+
 	return 0;
 }
 
@@ -654,6 +671,81 @@ static int ad4000_pwm_setup(struct spi_device *spi, struct ad4000_state *st)
 					st->cnv_trigger);
 }
 
+/*
+ * This executes a data sample transfer when using SPI offloading for when the
+ * device connections are in "3-wire" mode, selected when the adi,sdi-pin device
+ * tree property is absent or set to "high". In this connection mode, the ADC
+ * SDI pin is connected to MOSI or to VIO and ADC CNV pin is connected to a SPI
+ * controller CS (it can't be connected to a GPIO).
+ *
+ * In order to achieve the maximum sample rate, we only do one transfer per
+ * SPI offload trigger. This has the effect that the first sample data is not
+ * valid because it is reading the previous conversion result. We also use
+ * bits_per_word to ensure the minimum of SCLK cycles are used. And a delay is
+ * added to make sure we meet the minium quite time before releasing the CS
+ * line. Plus the CS change delay is set to ensure that we meet the minimum
+ * quiet time before asserting CS again.
+ * 
+ * This timing is only valid if turbo mode is enabled (reading during conversion).
+ */
+static int ad4000_prepare_offload_turbo_message(struct ad4000_state *st,
+						const struct iio_chan_spec *chan)
+{
+	struct spi_transfer *xfers = st->offload_xfers;
+
+	/* Have to do a short CS toggle to trigger conversion. */
+	xfers[0].cs_change = 1;
+	xfers[0].cs_change_delay.value = AD4000_TQUIET1_NS;
+	xfers[0].cs_change_delay.unit = SPI_DELAY_UNIT_NSECS;
+
+	/* HACK: invalid pointer indicates read data from SPI offload DMA */
+	xfers[1].rx_buf = (void*)(-1);
+	xfers[1].bits_per_word = chan->scan_type.realbits;
+	xfers[1].len = chan->scan_type.realbits > 16 ? 4 : 2;
+	xfers[1].delay.value = AD4000_TQUIET2_NS;
+	xfers[1].delay.unit = SPI_DELAY_UNIT_NSECS;
+
+	spi_message_init_with_transfers(&st->offload_msg, xfers, 2);
+
+	return devm_spi_optimize_message(&st->spi->dev, st->spi, &st->offload_msg);
+}
+
+/*
+ * This executes a data sample transfer when using SPI offloading for when the
+ * device connections are in "3-wire" mode, selected when the adi,sdi-pin device
+ * tree property is absent or set to "high". In this connection mode, the ADC
+ * SDI pin is connected to MOSI or to VIO and ADC CNV pin is connected to a SPI
+ * controller CS (it can't be connected to a GPIO).
+ *
+ * In order to achieve the maximum sample rate, we only do one transfer per
+ * SPI offload trigger. This has the effect that the first sample data is not
+ * valid because it is reading the previous conversion result. We also use
+ * bits_per_word to ensure the minimum of SCLK cycles are used. And a delay is
+ * added to make sure we meet the minium quite time before releasing the CS
+ * line. Plus the CS change delay is set to ensure that we meet the minimum
+ * conversion time before asserting CS again.
+ * 
+ * This timing is only valid if turbo mode is disabled (reading during acquisition).
+ */
+static int ad4000_prepare_offload_message(struct ad4000_state *st,
+					  const struct iio_chan_spec *chan)
+{
+	struct spi_transfer *xfers = st->offload_xfers;
+
+	/* HACK: invalid pointer indicates read data from SPI offload DMA */
+	xfers[0].rx_buf = (void*)(-1);
+	xfers[0].bits_per_word = chan->scan_type.realbits;
+	xfers[0].len = chan->scan_type.realbits > 16 ? 4 : 2;
+	xfers[0].delay.value = AD4000_TQUIET2_NS;
+	xfers[0].delay.unit = SPI_DELAY_UNIT_NSECS;
+	xfers[0].cs_change_delay.value = AD4000_TCONV_NS;
+	xfers[0].cs_change_delay.unit = SPI_DELAY_UNIT_NSECS;
+
+	spi_message_init_with_transfers(&st->offload_msg, xfers, 1);
+
+	return devm_spi_optimize_message(&st->spi->dev, st->spi, &st->offload_msg);
+}
+
 /*
  * This executes a data sample transfer for when the device connections are
  * in "3-wire" mode, selected when the adi,sdi-pin device tree property is
@@ -681,8 +773,9 @@ static int ad4000_prepare_3wire_mode_message(struct ad4000_state *st,
 	xfers[0].cs_change_delay.unit = SPI_DELAY_UNIT_NSECS;
 
 	xfers[1].rx_buf = &st->scan.data;
-	xfers[1].bits_per_word = chan->scan_type.storagebits;
-	xfers[1].len = BITS_TO_BYTES(chan->scan_type.storagebits);
+	xfers[1].len = chan->scan_type.realbits > 16 ? 4 : 2;
+	if (chan->scan_type.endianness != IIO_BE)
+		xfers[1].bits_per_word = chan->scan_type.realbits;
 	xfers[1].delay.value = AD4000_TQUIET2_NS;
 	xfers[1].delay.unit = SPI_DELAY_UNIT_NSECS;
 
@@ -713,7 +806,9 @@ static int ad4000_prepare_4wire_mode_message(struct ad4000_state *st,
 	xfers[0].delay.unit = SPI_DELAY_UNIT_NSECS;
 
 	xfers[1].rx_buf = &st->scan.data;
-	xfers[1].len = BITS_TO_BYTES(chan->scan_type.storagebits);
+	xfers[1].len = chan->scan_type.realbits > 16 ? 4 : 2;
+	if (chan->scan_type.endianness != IIO_BE)
+		xfers[1].bits_per_word = chan->scan_type.realbits;
 
 	spi_message_init_with_transfers(&st->msg, st->xfers, 2);
 
@@ -774,14 +869,12 @@ static int ad4000_probe(struct spi_device *spi)
 
 	/* Default to usual SPI connections if pin properties are not present */
 	st->sdi_pin = ret == -EINVAL ? AD4000_SDI_MOSI : ret;
+	bool using_spi_offload = spi_engine_ex_offload_supported(spi);
+
+	dev_info(dev, "Using %s connection mode\n", ad4000_sdi_pin[st->sdi_pin]);
+	dev_info(dev, "Using offload: %d\n", using_spi_offload);
 	switch (st->sdi_pin) {
 	case AD4000_SDI_MOSI:
-		indio_dev->info = &ad4000_reg_access_info;
-		if (spi_engine_ex_offload_supported(spi))
-			indio_dev->channels = &chip->reg_access_offload_chan_spec;
-		else
-			indio_dev->channels = &chip->reg_access_chan_spec;
-
 		/*
 		 * In "3-wire mode", the ADC SDI line must be kept high when
 		 * data is not being clocked out of the controller.
@@ -792,6 +885,18 @@ static int ad4000_probe(struct spi_device *spi)
 		if (ret < 0)
 			return ret;
 
+		indio_dev->info = &ad4000_reg_access_info;
+
+		if (using_spi_offload) {
+			indio_dev->channels = &chip->reg_access_offload_chan_spec;
+
+			ret = ad4000_prepare_offload_turbo_message(st, indio_dev->channels);
+			if (ret)
+				return ret;
+		} else {
+			indio_dev->channels = &chip->reg_access_chan_spec;
+		}
+
 		ret = ad4000_prepare_3wire_mode_message(st, indio_dev->channels);
 		if (ret)
 			return ret;
@@ -802,9 +907,13 @@ static int ad4000_probe(struct spi_device *spi)
 
 		break;
 	case AD4000_SDI_VIO:
-		if (spi_engine_ex_offload_supported(spi)) {
+		if (using_spi_offload) {
 			indio_dev->info = &ad4000_offload_info;
 			indio_dev->channels = &chip->offload_chan_spec;
+
+			ret = ad4000_prepare_offload_message(st, indio_dev->channels);
+			if (ret)
+				return ret;
 		} else {
 			indio_dev->info = &ad4000_info;
 			indio_dev->channels = &chip->chan_spec;
@@ -816,13 +925,12 @@ static int ad4000_probe(struct spi_device *spi)
 
 		break;
 	case AD4000_SDI_CS:
-		if (spi_engine_ex_offload_supported(spi)) {
-			indio_dev->info = &ad4000_offload_info;
-			indio_dev->channels = &chip->offload_chan_spec;
-		} else {
-			indio_dev->info = &ad4000_info;
-			indio_dev->channels = &chip->chan_spec;
-		}
+		if (using_spi_offload)
+			return dev_err_probe(dev, -EINVAL,
+					     "Cannot use SPI offload with 4-wire mode\n");
+
+		indio_dev->info = &ad4000_info;
+		indio_dev->channels = &chip->chan_spec;
 
 		ret = ad4000_prepare_4wire_mode_message(st, indio_dev->channels);
 		if (ret)
@@ -860,7 +968,7 @@ static int ad4000_probe(struct spi_device *spi)
 
 	ad4000_fill_scale_tbl(st, indio_dev->channels);
 
-	if (spi_engine_ex_offload_supported(spi)) {
+	if (using_spi_offload) {
 		ret = ad4000_pwm_setup(spi, st);
 		if (ret)
 			dev_err_probe(dev, ret, "PWM setup failed\n");
```

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -0,0 +1,942 @@
+	indio_dev->name = chip->dev_name;
+	indio_dev->num_channels = 1;
+
+	devm_mutex_init(dev, &st->lock);
```
Need to backport https://lore.kernel.org/linux-iio/20241030162013.2100253-2-andriy.shevchenko@linux.intel.com/

machschmitt commented on drivers/iio/adc/ad4000.c:
```diff
@@ -433,6 +545,15 @@ static int ad4000_write_raw(struct iio_dev *indio_dev,
 		}
 		unreachable();
+	case IIO_CHAN_INFO_SAMP_FREQ:
+		if (!val || val > st->max_rate_hz)
```
ah, yeah, this just reminds me I need to write to the config reg again to set the turbo mode. From the driver, we don't have much control over the actual VIO voltage. It can be understated in dt (or even absent) and above 2.7V in hw. I wouldn't limit the sample rate based on dt info. Don't know what upstream maintainers would think about it though. Not sure about the SCLK either. If SCLK < 80MHz we could be sampling less than 2MSPS and the sysfs interface stating we're at 2MSPS. That doesn't look good, but it's not usual to check for those parameters. Think we would be reviewing a lot of drivers if want to apply such checks consistently for IIO. For now, I'll take a lazy approach and not change it (unless required :) ).

machschmitt commented on drivers/iio/adc/ad4000.c:
```diff
@@ -611,16 +802,28 @@ static int ad4000_probe(struct spi_device *spi)
 	case AD4000_SDI_CS:
-		indio_dev->info = &ad4000_info;
-		indio_dev->channels = &chip->chan_spec;
+		if (spi_engine_ex_offload_supported(spi)) {
```
Good point. Will error on SDI_CS + offload configuration.

machschmitt commented on drivers/iio/adc/ad4000.c:
```diff
@@ -49,20 +54,22 @@
 		.sign = _sign,							\
 		.realbits = _real_bits,						\
 		.storagebits = _storage_bits,					\
 		.shift = _storage_bits - _real_bits,				\
```
ok, will try it.

machschmitt commented on drivers/iio/adc/ad4000.c:
```diff
@@ -463,10 +584,76 @@ static const struct iio_info ad4000_reg_access_info = {
+	if (IS_ERR(st->cnv_trigger))
+		return PTR_ERR(st->cnv_trigger);
+
+	ret = ad4000_set_sampling_freq(st, st->max_rate_hz);
```
hmm, okay, I'll have to do the reverse math of `ad4000_set_sampling_freq()` to get to the frequency from the PWM parameters. Another option would be to make the initial default samp freq proportional to the max one e.g. `ret = ad4000_set_sampling_freq(st, mult_frac(st->max_rate_hz, 3, 4);`.

machschmitt commented:
Hi @dlech thanks for diving into this and for the many suggestions to this driver.
I've applied all of your suggestions and a few additional changes I figured out were also needed while I was testing this.

Change log v2 -> v3
- Don't shift (0 shift) buffer data when using offload.
- Error on cs sdi-pin + offload config.
- Enable turbo mode if offload is supported.
- Applied changes suggested by @dlech.
- Use cs_hold to set offload non-turbo tCONV delay.
- Implemented timestamp channel.
- Fixed chan macros info_mask assignments.
- Lowered initial sample rate (I did it in a simplistic way for now since reversing the math at `ad4000_set_sampling_freq()` would take me some more time and I didn't want to take much longer to provide a reply to this).
- Removed `_raw` attribute for when offload is used (the invalid buffer pointer leads to kernel Oops on single-shot read).
- Picked up both patches that are in upstream mainling list.

dlech commented:
> * Removed `_raw` attribute for when offload is used (the invalid buffer pointer leads to kernel Oops on single-shot read).

With my changes, the `_raw` attribute was still working correctly since it was using a different SPI message. This is why I added a separate `offload_msg`, so that the `_raw` one could keep working too.

machschmitt commented:
> > * Removed `_raw` attribute for when offload is used (the invalid buffer pointer leads to kernel Oops on single-shot read).
> 
> With my changes, the `_raw` attribute was still working correctly since it was using a different SPI message. This is why I added a separate `offload_msg`, so that the `_raw` one could keep working too.

I see now why `ad4000_prepare_3wire_mode_message()` was called regardless of whether offload was supported or not.
Was expecting we would be able to reuse the same message for both offload and usual (non-offload) transfers. Turned out we do have to prepare specialized spi-transfer for each case. Anyways, reverted the removal of `_raw` attribute and made it always prepare the non-offload transfer as suggested. Thanks.

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -613,16 +907,34 @@ static int ad4000_probe(struct spi_device *spi)
+		} else {
+			indio_dev->info = &ad4000_info;
+			indio_dev->channels = &chip->chan_spec;
+			ret = ad4000_prepare_3wire_mode_message(st, indio_dev->channels);
```
I think we also need this one outside of the else similar to `case AD4000_SDI_MOSI`.

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -613,16 +907,34 @@ static int ad4000_probe(struct spi_device *spi)
-	case AD4000_SDI_CS:
 		indio_dev->info = &ad4000_info;
 		indio_dev->channels = &chip->chan_spec;
+
```
stray blank line added

dlech reviewed (commented):
Just a few more small things I noticed, but overall, I think this looks good now. :+1:

machschmitt commented:
okay, made the lastly suggested changes to the driver.
Oh, I forgot to mention I have included a small fix to `Documentation/iio/index.rst` at the beginning of the series.

Change log v3 -> v4:
- Always call `ad4000_prepare_3wire_mode_message()` at `AD4000_SDI_VIO` case so we can have single-shot return something that is not an error when setup is configured for offloading.
- Don't provide the soft timestamp channel for when offload is used.
- Removed extra blank line.

machschmitt commented:
Hi,

Any additional comments? @dlech ?
If this is good, can we merge it? @nunojsa

dlech commented on drivers/iio/adc/ad4000.c:
```diff
@@ -465,10 +600,152 @@ static const struct iio_info ad4000_reg_access_info = {
+	struct ad4000_state *st = iio_priv(indio_dev);
+	int ret;
+
+	spi_bus_lock(st->spi->master);
```
I would drop the `spi_bus_lock()`. The buffer can be enabled indefinitely, so this could cause a deadlock since it isn't an interruptible mutex. It's unlikely that there would be anything else on the same SPI bus anyway.

As it is though, we aren't releasing the lock on the error returns.

machschmitt replied:
done

dlech reviewed (commented):
LGTM other that one more small thing I noticed.
